### PR TITLE
Eliminate unnecessary numeric equality checks

### PR DESCRIPTION
**Pull request by nhunzaker**
This commit changes the way numeric equality for number inputs works such that it compares loose equality for numbers. This seems sufficient, and eliminates quite a bit of branching around numeric equality.

This additionally addresses Flow type issues on #11741, and it is based off that branch

**Test plan:**

1. Open http://react-concise-numbers.surge.sh
2. Run through the Text and Number test fixtures

**Tested in:**

* [x] Chrome 43, 62
* [x] Safari 7.1, 11
* [x] Firefox 52 (ESR), 58
* [x] IE 9, 10, 11
* [x] Edge 14, 16

### CODE REVIEW

**Comment by gaearon** [comment]
Can you explain why the original code was written this way, and why it's no longer necessary despite comments saying something about IE9?

File: packages/react-dom/src/client/ReactDOMFiberInput.js [hunk 1]
```diff
@@ -181,19 +181,9 @@ export function updateWrapper(element: Element, props: Object) {
...
-
-      if (
-        // eslint-disable-next-line
-        value != valueAsNumber ||
-        // eslint-disable-next-line
-        (value == valueAsNumber && node.value != value)
-      ) {
+    if (props.type === 'number') {
+      // eslint-disable-next-line
+      if (node.valueAsNumber !== value && node.value != value) {
```
Comment (nhunzaker) [inline packages/react-dom/src/client/ReactDOMFiberInput.js]: Actually I wonder if I even need to check valueAsNumber. Everything works as expected in IE9, which does not support this API.

File: packages/react-dom/src/client/ReactDOMFiberInput.js [hunk 2]
```diff
@@ -181,19 +181,9 @@ export function updateWrapper(element: Element, props: Object) {
...
-
-      if (
-        // eslint-disable-next-line
-        value != valueAsNumber ||
-        // eslint-disable-next-line
-        (value == valueAsNumber && node.value != value)
-      ) {
+    if (props.type === 'number') {
+      // eslint-disable-next-line
+      if (node.value != value) {
```
Comment (nhunzaker) [inline packages/react-dom/src/client/ReactDOMFiberInput.js]: This seems to be too good to be true, but this checks out in every browser. Unit tests also confirm it. Have I missed something?

Comment (gaearon) [reply packages/react-dom/src/client/ReactDOMFiberInput.js]: Worth checking git blame to see why this was introduced.

File: packages/react-dom/src/client/ReactDOMFiberInput.js [hunk 3]
```diff
@@ -181,19 +181,9 @@ export function updateWrapper(element: Element, props: Object) {
...
-
-      if (
-        // eslint-disable-next-line
-        value != valueAsNumber ||
-        // eslint-disable-next-line
-        (value == valueAsNumber && node.value != value)
-      ) {
+    if (props.type === 'number') {
+      // eslint-disable-next-line
+      if (node.value != value || (value === 0 && node.value === '')) {
```
Comment (nhunzaker) [inline packages/react-dom/src/client/ReactDOMFiberInput.js]: Looks like we didn't cover `""` to `0` in our unit tests. Added.

**Comment by nhunzaker** [comment]
> Can you explain why the original code was written this way [...]

Identifying all of the edge cases was a long road, and we didn't have all of the test fixtures and unit tests to cover edge cases in type coercion. It was the result of adding additional branches to cover different cases.

> [...] and why it's no longer necessary despite comments saying something about IE9?

IE9 does not support `input.valueAsNumber`, so we ran parseFloat on the value. However it looks like this is not necessary. The comment about IE9 justifies why we have to do that.

I think the gist of this is just that we do not need to parse the value as a number. We can make the following reductions (or at least this is my logic so we can try to poke holes in it):

```javascript
if (props.type === 'number') {
     // Simulate `input.valueAsNumber`. IE9 does not support it
     var valueAsNumber = parseFloat(node.value) || 0;
     if (
       // eslint-disable-next-line
       value != valueAsNumber ||
       // eslint-disable-next-line
       (value == valueAsNumber && node.value != value)
     ) {
```

`value == valueAsNumber` after doing a check for `value != valueAsNumber` is extraneous. We can collapse the check into:

```javascript
if (props.type === 'number') {
     // Simulate `input.valueAsNumber`. IE9 does not support it
     var valueAsNumber = parseFloat(node.value) || 0;
     if (
       // eslint-disable-next-line
       value != valueAsNumber && node.value != value
     ) {
```

To the best of my ability to test, `==` and `!=` have the same effect on a string as parseFloat. That means we can remove it:

```javascript
if (props.type === 'number') {
     // eslint-disable-next-line
     if (value != node.value && node.value != value) {
```

This is the same check! So finally we get to:

```javascript
if (props.type === 'number') {
     // eslint-disable-next-line
     if (value != node.value) {
```

The edge case here is that `''` loose equals `0`. That is controlled for in a branch earlier in the function, but we can collapse them to the same line:

```javascript
if (props.type === 'number') {
     // eslint-disable-next-line
     if (if (node.value != value || (value === 0 && node.value === '')) {
```

File: packages/react-dom/src/__tests__/ReactDOMInput-test.js [hunk 4]
```diff
@@ -248,6 +248,23 @@ describe('ReactDOMInput', () => {
     }
   });
 
+  it('performs as state change from "" to 0', () => {
```
Comment (aweary) [inline packages/react-dom/src/__tests__/ReactDOMInput-test.js]: "performs *a state change*

File: packages/react-dom/src/client/ReactDOMFiberInput.js [hunk 5]
```diff
@@ -181,19 +181,9 @@ export function updateWrapper(element: Element, props: Object) {
...
-      var valueAsNumber = parseFloat(node.value) || 0;
-
-      if (
-        // eslint-disable-next-line
-        value != valueAsNumber ||
-        // eslint-disable-next-line
-        (value == valueAsNumber && node.value != value)
-      ) {
+    if (props.type === 'number') {
+      // eslint-disable-next-line
```
Comment (aweary) [inline packages/react-dom/src/client/ReactDOMFiberInput.js]: Can we update this so it's only disabling the coercion rule?

File: packages/react-dom/src/client/ReactDOMFiberInput.js (hunk 3 above)
Comment (aweary) [inline packages/react-dom/src/client/ReactDOMFiberInput.js]: Before this change the `0`/ `""` coercion would occur for any type of input. Now it only applies to number inputs. Any chance that breaks something weird?

Comment (nhunzaker) [reply packages/react-dom/src/client/ReactDOMFiberInput.js]: Not from my ability to test. The strict equality check covers it.

File: packages/react-dom/src/__tests__/ReactDOMInput-test.js (hunk 4 above)
Comment (nhunzaker) [reply packages/react-dom/src/__tests__/ReactDOMInput-test.js]: Got it in 2ba3fe8e57f3d9c1cfd486d9c2e61a61c23404fb

File: packages/react-dom/src/client/ReactDOMFiberInput.js (hunk 5 above)
Comment (nhunzaker) [reply packages/react-dom/src/client/ReactDOMFiberInput.js]: Great idea. I also moved the loose check after the `value === 0` guard to avoid some extra DOM access. 3cdde77325a3f372298ac9084ff49041e817715f

**Comment by nhunzaker** [comment]
I'll check this into #11741 in a bit and that one should be good to go.
